File: Python3/Hard/WrongAnswer/minimum-time-to-visit-a-cell-in-a-grid_4.py
class Solution:
    def minimumTime(self, grid: List[List[int]]) -> int:
        
        n = len(grid)
        m = len(grid[0])
        queue = [(0, 0, grid[0][0])]
        visited = set()
        while queue:
            time, x, y = heapq.heappop(queue)
            if x == n - 1 and y == m - 1:
                return time
            for dx, dy in [(0, 1), (1, 0), (0, -1), (-1, 0)]:
                nx, ny = x + dx, y + dy
                if 0 <= nx < n and 0 <= ny < m and (nx, ny) not in visited:
                    heapq.heappush(queue, (max(time, grid[nx][ny]), nx, ny))
                    visited.add((nx, ny))
        return -1

File: Rust/Hard/WrongAnswer/minimum-time-to-visit-a-cell-in-a-grid_4.rs
impl Solution {
    pub fn minimum_time(grid: Vec<Vec<i32>>) -> i32 {
        
        let m = grid.len();
        let n = grid[0].len();
        let mut q = std::collections::BinaryHeap::new();
        let mut visited = vec![vec![false; n]; m];
        let mut res = 0;
        q.push(std::cmp::Reverse((0, 0, 0)));
        while !q.is_empty() {
            let (t, x, y) = q.pop().unwrap().0;
            if x == m - 1 && y == n - 1 {
                res = t;
                break;
            }
            if visited[x][y] {
                continue;
            }
            visited[x][y] = true;
            if x > 0 {
                q.push(std::cmp::Reverse((std::cmp::max(t, grid[x][y]), x - 1, y)));
            }
            if x < m - 1 {
                q.push(std::cmp::Reverse((std::cmp::max(t, grid[x][y]), x + 1, y)));
            }
            if y > 0 {
                q.push(std::cmp::Reverse((std::cmp::max(t, grid[x][y]), x, y - 1)));
            }
            if y < n - 1 {
                q.push(std::cmp::Reverse((std::cmp::max(t, grid[x][y]), x, y + 1)));
            }
        }
        if visited[m - 1][n - 1] {
            res
        } else {
            -1
        }
    }
}